gordon/metrics/log.py

```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+# TODO: update plugin loader to load metrics logger as default
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class LogTimer:
+    """Timer that outputs metrics to the logger.
+
+    Args:
+        metric (dict): Metric to output.
+        logger (.MetricLogger): Object used for outputting metrics.
+        time_unit (number): (optional) Multiplier used to scale time.time()
+            output.
+    """
+    def __init__(self, metric, logger, time_unit=1):
+        self.metric = metric
+        self.logger = logger
+        self.time_unit = time_unit
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        """Start the timer.
+        """
+        self._start_time = time.time()
+
+    async def stop(self):
+        """Stop the timer and output the time delta to the logger.
+        """
+        time_elapsed = (time.time() - self._start_time) * self.time_unit
+        self.metric['value'] = time_elapsed
+        self.logger.log(self.metric)
+
+
+class MetricLogger:
+    """Configurable wrapper that formats and outputs metrics.
+
+    Args:
+        level (str): Log level at which to output metric.
+        precision (str): Used to round metric to desired number of
+            places.
+    """
+    LOGFMT = '[{metric_name}]\tvalue: {value:.<prec>f}\tattrs: {attrs}'
```

Comment:
Any particular reason to tab-space?

Reply:
Also if `attrs` are optional, maybe not have them on the format string unless it's there?

Reply:
@econchick 

> Any particular reason to tab-space?

It's a shot at upping the readability, the image I'm going for is:

```bash
[request-acked]	value: 0.52	attrs: {'source': 'internal'}
[messages]	value: 132523	attrs: {'type': 'event'}
```

Compared to no-tabs:

```
```bash
[requests-latency] value: 0.52 attrs: {'source': 'internal'}
[messages-in-flight] value: 132523 attrs: {'type': 'event'}
```

Thoughts?

Reply:
I tend to prefer the spaces.  Tabs make you think something is gonna be organized, but it usually not how it works out. I think it usually ends up coming out like the following.

```
[dnsprovider-request-sent]  value: 132523   attrs:  {'source': 'internal'}
[cpu]   value:500   attrs: None
[messages-in-flight]    value: 0.52 attrs: {'type': 'event'}
```

Reply:
Thanks for the good example @descent3me - I got rid of the tabs.